File: Python/Medium/graphs/minScorePathBetweenCities.py
import collections
class Solution:
    def minScore(self, n: int, roads) -> int:
        graph = collections.defaultdict(dict)
        for n1, n2, distance in roads:
            graph[n1][n2]=distance
            graph[n2][n1] = distance

        min_val = float('inf')
        def DFS(node, visited):
            nonlocal min_val
            if node in visited:
                return
            
            visited.add(node)
            for neighbor in graph[node]:
                distance = graph[node][neighbor]
                if distance < min_val:
                    min_val = distance
                print(min_val)
                DFS(neighbor, visited)


        visited = set()
        DFS(1, visited)
        return min_val
    


n = 4
roads = [[1,2,9],[2,3,6],[2,4,7],[1,4,5]]

print(Solution().minScore(n, roads))

        

File: JavaScript/Medium/graphs/minScorePathBetweenCities.js
var minPath = function(n, roads) {
    let graph = {}

    for (const [n1,n2, distance] of roads){
        if (!graph[n1]) graph[n1] = {}
        if (!graph[n2]) graph[n2] = {}
        graph[n1][n2] = distance
        graph[n2][n1] = distance
    }

    var DFS = (node, visited ) => {
        if (!visited.has(node)) {
            return
        }

        visited.add(node)

        for (const neighbor of Object.keys(graph[node])){
            let distance = graph[node][neighbor]
            if (distance < minVal) minVal = distance
            DFS(neighbor, visited)
        }
    }

    let visited = new Set()
    let minVal = Infinity
    DFS(1, visited)

    return minVal
}